EventGenerator/src/GammaConversion_module.cc

```diff
@@ -0,0 +1,135 @@
+
+#include <iostream>
+#include <string>
+#include <cmath>
+#include <memory>
+
+#include "CLHEP/Vector/LorentzVector.h"
+#include "CLHEP/Random/RandFlat.h"
+
+#include "fhiclcpp/types/Atom.h"
+
+#include "art/Framework/Core/EDProducer.h"
+#include "art/Framework/Principal/Event.h"
+#include "art/Framework/Services/Registry/ServiceHandle.h"
+
+#include "Offline/SeedService/inc/SeedService.hh"
+#include "Offline/DataProducts/inc/PDGCode.hh"
+#include "Offline/MCDataProducts/inc/ProcessCode.hh"
+#include "Offline/MCDataProducts/inc/StageParticle.hh"
+#include "Offline/Mu2eUtilities/inc/simParticleList.hh"
+#include "Offline/Mu2eUtilities/inc/GammaPairConversionSpectrum.hh"
+
+namespace mu2e {
+
+  //================================================================
+  class GammaConversion : public art::EDProducer {
+  public:
+    struct Config {
+      using Name=fhicl::Name;
+      using Comment=fhicl::Comment;
+      fhicl::Atom<art::InputTag> inputSimParticles{Name("inputSimParticles"),Comment("A SimParticleCollection with input converted photons.")};
+      fhicl::Atom<int> material{Name("material"), Comment("Conversion material Z"), 13};
+      fhicl::Atom<bool> correlateAnglesOverKE{Name("correlateAnglesOverKE"), Comment("Option to correlate angles over KE"), true};
+      fhicl::Atom<int> verbosity{Name("verbosity"), Comment("Verbosity level"), 0};
+    };
+
+    using Parameters= art::EDProducer::Table<Config>;
+    explicit GammaConversion(const Parameters& conf);
+
+    virtual void produce(art::Event& event) override;
+
+    //----------------------------------------------------------------
+  private:
+
+    art::ProductToken<SimParticleCollection> const simsToken_;
+    const int material_;
+    const int verbosity_;
+
+    art::RandomNumberGenerator::base_engine_t& eng_;
+    CLHEP::RandFlat randFlat_;
+    GammaPairConversionSpectrum spectrum_;
+    GammaPairConversionSpectrum::elementData element_;
+    const ProcessCode process_;
+  };
+
+  //================================================================
+  GammaConversion::GammaConversion(const Parameters& conf)
+    : EDProducer{conf}
+    , simsToken_{consumes<SimParticleCollection>(conf().inputSimParticles())}
+    , material_{conf().material()}
+    , verbosity_{conf().verbosity()}
+    , eng_{createEngine(art::ServiceHandle<SeedService>()->getSeed())}
+    , randFlat_{eng_}
+    , spectrum_(&randFlat_, conf().correlateAnglesOverKE())
+    , element_(spectrum_.GetElementMap()[material_])
+    , process_(ProcessCode::mu2eGammaConversion)
+  {
+    produces<mu2e::StageParticleCollection>();
+  }
+
+  //================================================================
+  void GammaConversion::produce(art::Event& event) {
+    auto output{std::make_unique<StageParticleCollection>()};
+
+    const auto simh = event.getValidHandle<SimParticleCollection>(simsToken_);
+    const auto gammas = simParticleList(simh,PDGCode::gamma,ProcessCode::conv);
```

Comment:
How is this input SimParticle collection created? Is it from standard G4 processes? or do you assume a dedicated module for that (like RMC Generator)? Some description of the requirements on that collection would be helpful. For instance, how do you insure that the output doesn't double-count the conversion?

Reply:
This collection is created by selecting events with photon conversions using the StoppedParticlesFinder module and saving these conversion events. The GammaConversionPoints filter also only accepts photon conversions, specifically requiring them to be from the process of interest and within a given detector region (r in [0, 500] mm and z in [5000, 8000] mm for the main background estimates). I can add some description in the module that describes the expected input collection. From what I've seen the StoppedParticlesFinder output collection does not include the e+/e- pair from the photon conversion, but I'll try to look closer to ensure this is the case.